### PR TITLE
Add benchmarks comparing to Cairo VM

### PR DESCRIPTION
**Pull request by JulianGCalderon**
This PR benchmarks cairo native against cairo vm with criterion:
- Benchmarks cairo VM with fibonacci, factorial, and logistic map programs.
- Groups the benchmarks by program

Compile time is not taken into account for cairo native runs.

## Benchmarking Server

Factorial:
```
factorial_2M/Cached JIT time:   [290.86 ms 291.00 ms 291.15 ms]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

factorial_2M/Cached AOT time:   [904.06 ms 906.65 ms 909.17 ms]

factorial_2M/VM         time:   [3.7880 s 3.8034 s 3.8197 s]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```

Fibonacci:
```
fibonacci_2M/Cached JIT time:   [6.8784 ms 6.8821 ms 6.8864 ms]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high severe

fibonacci_2M/Cached AOT time:   [18.645 ms 18.665 ms 18.688 ms]
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low severe
  11 (11.00%) high severe

fibonacci_2M/VM         time:   [3.4395 s 3.4521 s 3.4659 s]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe
```

Logistic Map:
```
logistic_map/Cached JIT time:   [45.216 ms 45.244 ms 45.274 ms]
Found 13 outliers among 100 measurements (13.00%)
  8 (8.00%) high mild
  5 (5.00%) high severe

logistic_map/Cached AOT time:   [163.22 ms 163.48 ms 163.76 ms]
Found 16 outliers among 100 measurements (16.00%)
  16 (16.00%) high mild

logistic_map/VM         time:   [37.747 ms 37.778 ms 37.811 ms]
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe
```

## Macbook Air M1

```
factorial_2M/Cached JIT
                        time:   [271.52 ms 272.40 ms 274.08 ms]
factorial_2M/Cached AOT
                        time:   [734.26 ms 736.76 ms 740.13 ms]
factorial_2M/VM
                        time:   [3.9340 s 3.9436 s 3.9551 s]

fibonacci_2M/Cached JIT
                        time:   [7.6131 ms 7.6222 ms 7.6349 ms]
fibonacci_2M/Cached AOT
                        time:   [17.554 ms 17.562 ms 17.571 ms]
fibonacci_2M/VM
                        time:   [3.7504 s 3.7533 s 3.7565 s]

logistic_map/Cached JIT
                        time:   [45.113 ms 45.124 ms 45.136 ms]
logistic_map/Cached AOT
                        time:   [126.11 ms 126.17 ms 126.25 ms]
logistic_map/VM
                        time:   [42.685 ms 42.710 ms 42.735 ms]
```

### CODE REVIEW

File: benches/benches.rs [hunk 1]
```diff
@@ -36,25 +39,95 @@ fn criterion_benchmark(c: &mut Criterion) {
...
+        });
+
+        logistic_map_group.bench_function("VM", |b| {
+            b.iter(|| {
+                logistic_map_runner.run_function_with_starknet_context(
+                    logistic_map_function,
+                    &[],
+                    Some(usize::MAX),
+                    StarknetState::default(),
+                )
```
Comment (Oppen) [inline benches/benches.rs:125]: The VM is stateful, so you should probably clone the runner before execution. There are a few methods in criterion that allow that kind of setup.

Comment (Oppen) [reply benches/benches.rs:125]: This may or may not apply entirely given we're using the `cairo-lang` version rather than ours directly.

Comment (JulianGCalderon) [reply benches/benches.rs:125]: I reviewed `cairo-lang`'s runner and I don't think this applies, as a VM is created on each execution:

SierraCasmRunner: https://github.com/starkware-libs/cairo/blob/2203a47f8a098cd4718d03bd109ca014049419e7/crates/cairo-lang-runner/src/lib.rs#L162-L178
Run Method: https://github.com/starkware-libs/cairo/blob/2203a47f8a098cd4718d03bd109ca014049419e7/crates/cairo-lang-runner/src/lib.rs#L459-L460

File: benches/benches.rs [hunk 2]
```diff
@@ -36,25 +39,95 @@ fn criterion_benchmark(c: &mut Criterion) {
...
+        factorial_group.bench_function("Cached JIT", |b| {
+            b.iter(|| jit_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
+        });
+        factorial_group.bench_function("Cached AOT", |b| {
+            b.iter(|| aot_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
+        });
+
+        factorial_group.bench_function("VM", |b| {
+            b.iter(|| {
+                factorial_runner.run_function_with_starknet_context(
```
Comment (pefontana) [inline benches/benches.rs:68]: We need to find another way to execute the VM.
See the `fn run_function_with_starknet_context`
```
        let assembled_program = self.casm_program.clone().assemble_ex(&entry_code, &footer);
```
It clone the casm_program adding lot of overhead to the execution

We need to discuss how we do this, maybe we have to add a new method to the `cairo-vm/cairo1-run` crate

**Review by pefontana — CHANGES_REQUESTED** [review]

File: benches/benches.rs [hunk 3]
```diff
@@ -36,25 +39,95 @@ fn criterion_benchmark(c: &mut Criterion) {
...
+
+        logistic_map_group.bench_function("Cached AOT", |b| {
+            b.iter(|| {
+                aot_logistic_map.invoke_dynamic(logistic_map_function_id, &[], Some(u128::MAX))
+            });
+        });
+
+        logistic_map_group.bench_function("VM", |b| {
+            b.iter(|| {
+                logistic_map_runner.run_function_with_starknet_context(
```
Comment (pefontana) [inline benches/benches.rs:120]: A doubt I got seeing the benchmark numbers, are we sure this program is not failing?
We are not checking the result of the run_function_with_starknet_context call

**Comment by pefontana** [comment]
Other thing to take into account, maybe is easier to benchmark a contract execution

File: benches/benches.rs (hunk 3 above)
Comment (JulianGCalderon) [reply benches/benches.rs]: I've added some checks to assert this. It seems to be working properly.

File: benches/benches.rs (hunk 2 above)
Comment (JulianGCalderon) [reply benches/benches.rs]: I think that particular issue has been fixed: https://github.com/starkware-libs/cairo/blob/226c3646be247828c08b737a73f6da10de76e0fc/crates/cairo-lang-runner/src/lib.rs#L191
```rust
let assembled_cairo_program = self.casm_program.assemble_ex(&header, &footer);
```

**Pull request merged**